clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1096,96 +1097,156 @@ public void assign(Collection<TopicPartition> partitions) {
+        return poll(timeout, true);
+    }
+
+    private ConsumerRecords<K, V> poll(final Duration timeout, final boolean includeMetadataInTimeout) {
         acquireAndEnsureOpen();
         try {
-            if (timeout < 0)
-                throw new IllegalArgumentException("Timeout must not be negative");
+            if (timeout.isNegative()) throw new IllegalArgumentException("Timeout must not be negative");
 
-            if (this.subscriptions.hasNoSubscriptionOrUserAssignment())
+            if (this.subscriptions.hasNoSubscriptionOrUserAssignment()) {
                 throw new IllegalStateException("Consumer is not subscribed to any topics or assigned any partitions");
+            }
 
             // poll for new data until the timeout expires
-            long start = time.milliseconds();
-            long remaining = timeout;
+            long remainingMs = timeout.toMillis();
             do {
-                Map<TopicPartition, List<ConsumerRecord<K, V>>> records = pollOnce(remaining);
+
+                client.maybeTriggerWakeup();
+
+                if (includeMetadataInTimeout) {
+                    final long metadataStart = time.milliseconds();
+                    if (!updateAssignmentMetadataIfNeeded(Duration.ofMillis(remainingMs))) {
+                        return ConsumerRecords.empty();
+                    }
+                    final long metadataEnd = time.milliseconds();
+                    remainingMs = Math.max(0L, remainingMs - (metadataEnd - metadataStart));
+                } else {
+                    while (!updateAssignmentMetadataIfNeeded(Duration.ofMillis(Long.MAX_VALUE))) {
+                        log.warn("Still waiting for metadata");
+                    }
+                }
+
+                final long fetchStart = time.milliseconds();
+                final Map<TopicPartition, List<ConsumerRecord<K, V>>> records = pollForFetches(remainingMs);
+
                 if (!records.isEmpty()) {
                     // before returning the fetched records, we can send off the next round of fetches
                     // and avoid block waiting for their responses to enable pipelining while the user
                     // is handling the fetched records.
                     //
                     // NOTE: since the consumed position has already been updated, we must not allow
                     // wakeups or any other errors to be triggered prior to returning the fetched records.
-                    if (fetcher.sendFetches() > 0 || client.hasPendingRequests())
+                    if (fetcher.sendFetches() > 0 || client.hasPendingRequests()) {
                         client.pollNoWakeup();
+                    }
 
                     return this.interceptors.onConsume(new ConsumerRecords<>(records));
                 }
+                final long fetchEnd = time.milliseconds();
+                remainingMs = Math.max(0L, remainingMs - (fetchEnd - fetchStart));
```

Comment:
Yeah... This is a result of double-refactoring. I got away from the "elapsed" approach when I introduced the "remainingTimeAtLeastZero" method, and then I made a recent change to avoid calling system time so much. I would have gone back to the original time-tracking approach, but I'd forgotten about it.

You are right, tracking elapsed time doesn't rack up rounding errors. I'll add another patch momentarily using this approach throughout my diff instead of "remainingTime".